File: prime-number/sieve_prime_factorization.cpp
#include <bits/stdc++.h>
#include <cmath>
using namespace std;
const int len = 1001;

void Sieve_of_Eratosthenes(int n)
{
    // intialzie a bool array to store the value of is_prime
    bool num[n];
    memset(num, true, sizeof(num));
    for (int i = 2; i <= sqrt(n); i++)
    {
        if (num[i] == true)
        {
            for (int j = i * i; j <= n; j += i)
            {
                num[j] = false;
            }
        }
    }
    vector<int> prime;
    for (int i = 2; i <= n; i++)
    {
        if (num[i])
        {
            prime.push_back(i);
        }
    }
    for (int i = 0; i <= prime.size() - 1; i++)
    {
        cout << prime[i] << " ";
    }
}

void least_prime_factor(int n)
{
    // intialize a vector to store least prime
    vector<int> least_prime(n + 1, 0);
    for (int i = 2; i < n; i++)
    {
        // if i is prime
        if (least_prime[i] == 0)
        {
            // least prime factor of the prime number is itself
            least_prime[i] = i;
            for (int j = i * i; j <= n; j += i)
            {
                // to gurantee that prime factor of the number is least
                // eg, 6 = 2 * 3
                if (least_prime[j] == 0)
                {
                    least_prime[j] = i;
                }
            }
        }
    }
    for (int i = 2; i < n; i++)
    {
        cout << "least prime factor " << i << ": " << least_prime[i] << "\n";
    }
}

void sieve(int n)
{
    int spf[len];
    // initialize the array and set all of its elements to 0
    fill_n(spf, len, 0);
    for (int i = 2; i < len; i++)
    {
        if (spf[i] == 0)
        {
            spf[i] = i;
            for (int j = i * i; j < len; j += i)
            {
                if (spf[j] == 0)
                {
                    spf[j] = i;
                }
            }
        }
    }

    vector<int> arr;
    cout << "prime factorization for " << n << " : ";
    while (n != 1)
    {
        arr.push_back(spf[n]);
        n /= spf[n];
    }

    for (int i = 0; i < arr.size(); i++)
    {
        cout << arr[i] << " ";
    }
}

int main()
{
    int n = 100;
    Sieve_of_Eratosthenes(n);
    least_prime_factor(n);
    sieve(n);
    return 0;
}

File: prime-number/sieve_prime_factorization.nim
import math
import algorithm
const len=1000

proc Sieve_of_Eratosthenes(n: int)=
    # initialize the array and fill all its elements with true
    var a: array[len, bool]
    a.fill(true)

    for i in 2..int(sqrt(n.float)):
        if a[i]:
            for j in countup(i*i, n, i):
                if a[j]:
                    a[j]=false

    var prime: seq[int] = @[]
    for i in 2..n:
        if a[i]:
            prime.add(i)
    echo prime

proc least_prime_factor(n: int)=
    var a: array[len, int]
    a.fill(0)

    for i in countup(2, n):
        if a[i]==0:
            a[i]=i
            for j in countup(i*i, n, i):
                if a[j]==0:
                    a[j]=i

    var num: seq[int]
    for i in 2..n:
        num.add(i)
    echo "the least prime factor is"
    echo num
    echo a[2 .. n]

proc sieve(n: int)=
    var spf: array[len, int]
    spf.fill(0)

    for i in countup(2, n):
        if spf[i]==0:
            spf[i]=i
            for j in countup(i*i, n, i):
                if spf[j]==0:
                    spf[j]=i
    echo "the prime factors of ", n, ":"
    var factors: seq[int]
    var num=n
    while num != 1:
        factors.add(spf[num])
        num=num div spf[num]
    echo factors

var n=100
Sieve_of_Eratosthenes(n)
least_prime_factor(n)
sieve(n)